src/main/scala/Validator.scala

```diff
@@ -76,22 +147,170 @@ class Validator(var tableCsvFile: URI, sourceUri: String = "") {
             ""
           )
         }
+
         val result = table.validateRow(row)
         result match {
           case Some(validateRowOutput) => {
-            val primaryKeyValues = validateRowOutput.primaryKeyValues
-            if (allPrimaryKeyValues.contains(primaryKeyValues)) {
-              errors = errors :+ ErrorWithCsvContext(
-                "duplicate_key",
+            errors ++= validateRowOutput.warningsAndErrors.errors
+            warnings ++= validateRowOutput.warningsAndErrors.warnings
+            setChildTableForeignKeys(validateRowOutput, childTableForeignKeys)
+            setParentTableForeignKeyReferences(
+              validateRowOutput,
+              parentTableForeignKeyReferences
+            )
+            validatePrimaryKeys(allPrimaryKeyValues, row, validateRowOutput)
```

Comment:
Instead of validating at this stage, I'd move the row validation functionality out into a new function to make it easier to parallelise later. I'd ensure that instead of calling `validatePrimaryKeys`, we returned the (errorsAndWarnings, primaryKey, childTableForeignKeys, parentTableForeignKeyReferences) and let the caller bring them all together and check for uniqueness.

Reply:
This is a bit hard to explain, probably something we should have a call about.